### ISSUE TITLE
Trying to Fetch Ubuntu Will Failed If Language Set to zh-TW

### ISSUE THREAD

**Issue by wynn1212** [bedrocklinux/bedrocklinux-userland#110]
When language set to zh-TW. Fetching ubuntu will fail because there is no `language-pack-zh` only `language-pack-zh-hant` is available
```
[13/13 (100%)}] Configuring
Hit:1 http://mirrors.layeronline.com/ubuntu cosmic InRelease
Get:2 http://mirrors.layeronline.com/ubuntu cosmic/main Translation-en [513 kB]
Get:3 http://archive.ubuntu.com/ubuntu cosmic-security InRelease [88.7 kB]        
Get:4 http://archive.ubuntu.com/ubuntu cosmic-security/main amd64 Packages [143 kB]
Get:5 http://archive.ubuntu.com/ubuntu cosmic-security/main Translation-en [55.7 kB]
Fetched 801 kB in 3s (296 kB/s)         
Reading package lists... Done
Reading package lists... Done
Building dependency tree... Done
Package language-pack-zh is not available, but is referred to by another package.
This may mean that the package is missing, has been obsoleted, or
is only available from another source
However the following packages replace it:
  language-pack-zh-hant language-pack-zh-hans

E: Package 'language-pack-zh' has no installation candidate
ERROR: Unexpected error occurred.
root@bedrock:~# 
```

**Comment by paradigm** [comment]
If I understand correctly, it should use `language-pack-zh-hant` if it sees `zh_TW` and `language-pack-zh-hans` if it sees `zh_CN`, correct?

Try editing `/bedrock/share/brl-fetch/distros/ubuntu` and changing [the langpack line](https://github.com/bedrocklinux/bedrocklinux-userland/blob/aa629a430153cd43fb3468db2a77e4dfbcc3df91/src/slash-bedrock/share/brl-fetch/distros/ubuntu#L159) to

```
		case "${localegen}" in
			zh_CN) langpack="language-pack-zh-hans"
			zh_TW) langpack="language-pack-zh-hant"
			*) langpack="language-pack-$(echo "${localegen}" | sed 's/^\(..\).*$/\1/')" ;;
		esac
```

Then give that a try and let me know if it works for you.  If it does, I will probably make that change in a future update.

**Comment by wynn1212** [comment]
I don't know if I'm replace it right. But I'm getting the following error when preform `brl fetch ubuntu`

```
/bedrock/libexec/brl-fetch: /bedrock/share/brl-fetch/distros/ubuntu: line 161: syntax error: unexpected ")" (expecting ";;")
ERROR: Unexpected error occurred.
```
Here's a `/bedrock/share/brl-fetch/distros/ubuntu`
```
156         fi
157         localegen="$(cfg_value "locale" "localegen")"
158         if [ -n "${localegen:-}" ]; then
159                 case "${localegen}" in
160                         zh_CN) langpack="language-pack-zh-hans"
161                         zh_TW) langpack="language-pack-zh-hant"
162                         *) langpack="language-pack-$(echo "${localegen}" | sed 's/^\(..\).*$/\1/')" ;;
163                 esac
164                 yes "" | LANG=C chroot "${target_dir}" /usr/bin/apt-get update
165                 yes "" | LANG=C chroot "${target_dir}" /usr/bin/apt-get install "${langpack}"
```

**Comment by paradigm** [comment]
What you did was correct; I mistyped.

Try this:

```
		case "${localegen}" in
			zh_CN) langpack="language-pack-zh-hans" ;;
			zh_TW) langpack="language-pack-zh-hant" ;;
			*) langpack="language-pack-$(echo "${localegen}" | sed 's/^\(..\).*$/\1/')" ;;
		esac
```

Note the `;;` on two lines that I forgot previously.

**Comment by wynn1212** [comment]
Nope! same thing.

I don't know why ubuntu still trying to install `language-pack-zh` instead of `language-pack-zh-hant`

Here's configuration for bedrock.conf
https://gist.github.com/wynn1212/61db16447e8f4e777b50933d0cdc735b

Here's the output of `locale`:
```
LANG=zh_TW.UTF-8
LANGUAGE=zh_TW:zh
LC_CTYPE="zh_TW.UTF-8"
LC_NUMERIC="zh_TW.UTF-8"
LC_TIME="zh_TW.UTF-8"
LC_COLLATE="zh_TW.UTF-8"
LC_MONETARY="zh_TW.UTF-8"
LC_MESSAGES="zh_TW.UTF-8"
LC_PAPER="zh_TW.UTF-8"
LC_NAME="zh_TW.UTF-8"
LC_ADDRESS="zh_TW.UTF-8"
LC_TELEPHONE="zh_TW.UTF-8"
LC_MEASUREMENT="zh_TW.UTF-8"
LC_IDENTIFICATION="zh_TW.UTF-8"
LC_ALL=
```

**Comment by paradigm** [comment]
To just get you going,  let's just try removing the language package lines.  Delete these lines: https://github.com/bedrocklinux/bedrocklinux-userland/blob/aa629a430153cd43fb3468db2a77e4dfbcc3df91/src/slash-bedrock/share/brl-fetch/distros/ubuntu#L158-L164 and try again.  Once the fetch is finished, install the appropriate language package.

I'll mess with it more on my own to see why the above fix did not work.  I can try using your example `bedrock.conf` when testing it.  Leave this issue open - I'll close it once I have a working solution.

**Comment by wynn1212** [comment]
OK, lets try to recreate this situation again.
My initial system is `Debian Buster (Testing)` before install Bedrock.
When selecting language. I choose Chinese `(Traditional) - 中文(繁體).`
After installed my system and hijacked into bedrock linux, when I trying to fetch ubuntu. It will failed on installing language pack

Strangely enough. If I switch my initial system to `fedora 29` and language set to `(Traditional) - 中文(繁體).`  During fetching ubuntu. It'll just install `langpack-en` instead of `language-zh-hant`. Hence error did't occurred when fetching ubuntu in fedora.

**Comment by paradigm** [comment]
> OK, lets try to recreate this situation again.
> My initial system is `Debian Buster (Testing)` before install Bedrock.
> When selecting language. I choose `Chinese (Traditional) - 中文(繁體).`
> After installed my system and hijacked into bedrock linux, when I trying to fetch ubuntu. It will failed on installing language pack

I believe this is because Bedrock's logic to determine which Ubuntu language package to install is overly simple.  In my very limited exploration of this area, I found many `language-pack-<xx>` packages which correspond to `<xx>_<YY>.<ZZ>`, and so that is how Bedrock maps things.  I did not do enough research to know about `language-pack-<xx>-<yyyy>` packages such as `language-pack-zsh-hant`/`language-pack-zsh-hans`.  Now that I know about them, I can adjust things accordingly.  I will (eventually) release an update which makes this work.

For now, rather than waiting for the update, I think you can just remove the lines where Bedrock tries install the `language-pack` packages, and instead simply manually do so yourself after the fetch has completed.

> Strangely enough. If I switch my initial system to `fedora 29` and language set to `Chinese (Traditional) - 中文(繁體)`. During fetching ubuntu. It'll just install `langpack-en` instead of `language-zh-hant`. Hence error did't occurred when fetching ubuntu in fedora.

This is likely a separate issue in Bedrock where it is not detecting the language from the initial system correctly.  I'll look into this as well.  My guess is the Fedora system has both English _and_ Chinese language systems enabled, and Bedrock is only going with the first it sees.  I will have to make it smart enough to understand multiple options.

For now, you can edit `/bedrock/etc/bedrock.conf` and set the `localegen` and `LANG` items under `[locale]` to your liking.

**Comment by paradigm** [comment]
This issue should be resolved in `0.7.3`.  Please either `brl update` to it or hijack a fresh install when you get the chance, try to fetch Ubuntu again, then let me know if it works this time for you.

**Comment by wynn1212** [comment]
Yes, this time it works by NOT installing language pack in `0.7.3`. But why do this on purpose?

**Comment by wynn1212** [comment]
Thanks! It works correctly. Now closing the issue.
